### PR TITLE
[in progress] add op to mkldnn_convolution

### PR DESCRIPTION
**Pull request by george-qi**
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: George Qi <georgeqi@fb.com>

Fixes #{issue number}

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -7765,6 +7764,26 @@ def test_out_variant(self):
             with self.assertRaisesRegex(RuntimeError, "out= function"):
                 torch.add(foo, bar, out=bar)
 
+    def test_mkldnn_convolution(self):
+        n = 2
+        c = 3
+        o = 4
+        x = 5
+
+        inp = torch.rand(n, c, x, dtype=torch.float, requires_grad=True)
```
Comment (albanD) [inline test/test_autograd.py:7773]: I'm not sure your inputs are correct here. For a 2d conv, you want a 4d input, and 4d weights.
Also the last two dimensions of the input should not match the weights otherwise otherwise padding/stride will be ignored.

I feel like an input of shape (2, 2, 5, 5), weight (3, 2, 3, 3), padding (1, 1), dilation (1, 1) and groups 1 would be better

File: tools/autograd/templates/VariableType.cpp [hunk 2]
```diff
@@ -36,6 +36,9 @@ namespace torch { namespace autograd {
 
 namespace VariableType {
 namespace{
+  Tensor toNonOptZerosLikeTensor(const c10::optional<Tensor>& t) {
```
Comment (albanD) [inline tools/autograd/templates/VariableType.cpp:39]: We should eventually move this back. Can you add a TODO here if you don't have time to investigate?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -4184,6 +4184,46 @@ Tensor _lu_with_info_jvp(
...
+  const Tensor& self_p,
+  const Tensor& self_t,
+  const Tensor& weight_p,
+  const Tensor& weight_t,
+  const Tensor& bias_t,
+  const at::IntArrayRef padding,
+  const at::IntArrayRef stride,
+  const at::IntArrayRef dilation,
+  const int groups,
+  const Tensor& result
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:4197]: Why do you take result as input here?

File: test/test_autograd.py (hunk 1 above)
Comment (georgeqi94) [reply test/test_autograd.py:7773]: Just want to check - for weight, should it be (2, 2, 3, 3) since the first dimension should match input?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (georgeqi94) [reply torch/csrc/autograd/FunctionsManual.cpp:4197]: Good point - removed

File: test/test_autograd.py [hunk 4]
```diff
@@ -7765,6 +7764,24 @@ def test_out_variant(self):
...
+
+        padding = (2, 3)
+        stride = (1, 1)
+        dilation = (3, 1)
+        groups = 1
+
+        torch.autograd.gradcheck(
+            torch.mkldnn_convolution,
+            [inp, weight, bias, padding, stride, dilation, groups],
+            eps=1e-3,
```
Comment (georgeqi94) [inline test/test_autograd.py:7781]: Added 1e-3 because of this post - https://discuss.pytorch.org/t/why-is-nn-linear-failing-gradcheck-is-there-a-bug-in-gradcheck/17346/2

gradcheck uses double precision while `mkldnn_convolution` requires float or b16

Comment (albanD) [reply test/test_autograd.py:7781]: Yes, that's unfortunate...

File: test/test_autograd.py [hunk 5]
```diff
@@ -7765,6 +7764,24 @@ def test_out_variant(self):
             with self.assertRaisesRegex(RuntimeError, "out= function"):
                 torch.add(foo, bar, out=bar)
 
+    # TODO: Temporary solution to substitute for OpsInfo
+    def test_mkldnn_convolution(self):
+        inp = torch.rand(2, 2, 5, 5, dtype=torch.float, requires_grad=True)
+        weight = torch.rand(2, 2, 3, 3, dtype=torch.float, requires_grad=True)
+        bias = torch.rand(2, dtype=torch.float, requires_grad=True)
+
+        padding = (2, 3)
```
Comment (albanD) [inline test/test_autograd.py:7773]: I'm still not sure that padding is very representative. For a 3x3 kernel, people usually use (1, 1) no?

**Review by albanD — APPROVED** [review]
LGTM

We can improve testing when we merge all the conv implementation.
